r/R/dplyr-funcs-datetime.R

```diff
@@ -147,5 +147,16 @@ register_bindings_datetime <- function() {
   register_binding("pm", function(x) {
     !call_binding("am", x)
   })
+  register_binding("tz", function(x) {
+    if (!call_binding("is.POSIXct", x)) {
```

Comment:
This line isn't quite right, actually. We want to use something like `!is.POSIXct(x)` that will be compatible with either R objects or Arrow types. I know there is at least one place that we test if a thing is a timestamp like that. Could you see if you can find it + see what that uses to get this working?

That _should_ allow one to then `arrow:::call_binding("tz", "2020-10-01")` and get an error directly.

Reply:
Oh, actually no — I'm totally wrong this is fine it's this that's the problem: 

```
if (inherits(x, "Expression")) {
  class <- x$type()$ToString()
} else {
  class <- type(x)
}
```

That  should actually be:

```
if (inherits(x, "Expression")) {
  class <- x$type()$ToString()
} else {
  class <- type(x)$ToString()
}
```

and realistically, that should not even need an `if` there and could be `type(x)$ToString()` if we added (or fixed) a `type` method for expressions (I get `Error: Unrecognized vector instance for type ENVSXP` when I try `type({Expression})`) But that doesn't have to happen here (though we should make a Jira for it!

Reply:
https://issues.apache.org/jira/browse/ARROW-15712

Reply:
I have a PR (#12447) that adds a `type` method for `Expression` so we could rebase once that PR is merged.